src/encord_active/lib/file_structure/base.py

```diff
@@ -0,0 +1,79 @@
+from abc import abstractmethod
+from pathlib import Path
+from typing import Iterator, Optional, Union
+
+from prisma import models
+
+from encord_active.lib.db.base import DataUnitLike
+
+
+class BaseProjectFileStructure:
```

Comment:
Sure, in order to be able to use the database in `ProjectFileStructure` to iterate label rows and data units as we used to do via the filesystem, we need to avoid the cyclic dependency between `ProjectFileStructure` and `PrismaConnection`. That's the main reason why now we add an instance of `BaseProjectFileStructure` to the db connections instead of `project_path` (avoid referencing `ProjectFileStructure` in the db connection code while keeping the type hints and autocompletion). Also, as a collateral is the first step to enable customisable project file structures, not necessarily to enable it for the client but adding flexibility in future file structure changes if we manage to put the type of file structure in a config file.

To sum up:
1. Avoid cyclic dependencies.
2. Adds customisation goddies.

Reply:
I see, mostly makes sense although I'm personally not a fan of abstraction as a solution to a cyclic dependency, if anything it just hides it away.

If I may suggest another approach, I think composition would be preferable in this case.
Instead of moving the whole class into a base class, we could instead move the logic to iterate label rows and data units into its own class that also depends on `ProjectFileStructure` which seems like a more appropriate separation. wdyt?

Reply:
IMO more than an abstraction is an interface but we don't have those in python 😄. At some point we started adding logic non-related to the filesystem into `ProjectFileStructure`, so we definitely should separate that logic. Probably we can bear with this one for now, test the robustness of the logic and afterwards migrate it to its own component (while we completely get rid of the old and also not liked behaviour added to `ProjectFileStructure`).